### ISSUE TITLE
Support dynamic client registration

### ISSUE THREAD

**Issue by damooo** [sebadob/rauthy#165]
[RFC 2591](https://datatracker.ietf.org/doc/html/rfc7591) defines "OAuth 2.0 Dynamic Client Registration Protocol" (DCRM), which enables clients to register themselves directly at a standard endpoint.

It must be simple enough to create required endpoints, with just copying current client endpoints, and changing any auth details.

We can have the feature behind a config.

See also: [OAuth 2.0 Dynamic Client Registration Management (DCRM) Protocol](https://medium.com/@abilashini/oauth-2-0-dynamic-client-registration-management-dcrm-protocol-b55f222f481b)

I could make a PR, if feature is ok.

**Comment by sebadob** [comment]
I thought about this, but I don't see a proper use case for this to be honest.

I would never, ever, make this just open to anyone, because bots and spammers will just waste your database. To have a mechanism against this, we would need to allow registration with a Token only. However, to make this process somewhat secure, this Token should be rotated and it must be dynamically changing. If this is the case though, you would need an Admin for the instance to actually tell you the current token in some way. But the thing is, if you do this beforehand, the Admin could just register the client directly in the first place.

Rauthy does resolve ephemeral clients via URL if allowed, and I see a reason to have this feature for Solid OIDC. But I just don't see one for a real dynamic client registration.
Do you have an example where this actually would make sense and provide a real benefit?

**Comment by damooo** [comment]
In decentralized scenarios, applications just register at random servers, and work with resource servers. The solidos project uses them at through out.

Also, it is the only way for decentralized localhost apps, or private apps, as they can not be ephemereal clients as their uri will not be accesable to rauthy.

**Comment by sebadob** [comment]
As a server Admin, I would never want to have this feature enabled. This would be a nightmare to manage. You will loose control over the whole situation at some point and scale imo.

For these szenarios, you could just have a static, generic client that allows all possible redirects to any localhost instance. Anyone could just use this client and there is no need to register thousands of clients that all do the same thing, when you could have just one that allows all these redirects to localhost. I do this on a daily basis in local development already.

The problem is, that this would introduce a big overhead into Rauthy. You don't just need an endpoint for the registration, but all these clients would need to be able to manage and update themselves, and so on.

**Comment by damooo** [comment]
Can we have behind some config like `UNSAFE_DCR`? Keycloak also [support DCR](https://www.keycloak.org/docs/latest/securing_apps/#openid-connect-dynamic-client-registration). Don't know what measures they adopt.

**Comment by sebadob** [comment]
Yes sure we could implement this. This is just not something small and we should provide the option to at least specify an optinal static token via config for the registration.

The next steps for me would be now to complete the `webid` feature and then have another release most probably.

**Comment by sebadob** [comment]
We only need a good stretegy for the client ids. Like it is implemented now, you need to specify the id during registration.
I think in the case of dcr, they should be auto generated by rauthy and have an id like for instance `dyn-{16randomAlnumChars}` or something like that.

**Comment by damooo** [comment]
Are there anything i should pr for webid? Removing is_open, etc.?

**Comment by sebadob** [comment]
I need to re-work the UI part for the `webid`s aynway, since it broke with the latest PR.
At the same time I will do some other adjustments and move the endpoint URL.
But you could help with some testing of the UI part later on, when it is finished.

**Comment by sebadob** [comment]
> This would also be required if the native OIDC approach would be supported without the Matrix Authentication Service.

No it's not required. I am using synapse + Rauthy without MAS in between since a few months now,

Currently, dynamic client registration does not really make much sense to me, considering all the bots and spammers these days.

**Comment by Tremoneck** [comment]
The new oidc native flow is explained in more detail [here](https://areweoidcyet.com/client-implementation-guide/). Are you using the native oidc flow or the default legacy one? 
In the old way synapse is the client and the resource owner. In the future Element/Fluffy hat etc. is the client and synapse only the resource owner.

**Comment by sebadob** [comment]
I am using the now called "legacy" approach, but:

> "... each client either needs to be pre-registered with a HS/OP or the HS/OP allows clients to dynamically register ..."

According to that, both ways work fine.

**Comment by damooo** [comment]
It may be made optional, so that admins can opt-in explicitely to enable newer  usecases.   May be few best practice criterion for managing the spam can be introduced.

**Comment by Tremoneck** [comment]
The static registration would only work if one provides the client with the secret build in. Like a self hosted element web, or by distributing a self compiled app. Otherwise you get the problem of the user having to register his client first with secret and everything, if the client even has an interface supporting this. Without support the new flow can be used, with the layer of MAS, that is it's purpose after all.

It would need to be explicitly opt in for the server admin. A solution against spammers would be to only persist the registered client to the dB, after it has been used by a user.

**Comment by sebadob** [comment]
As commented [here](https://github.com/sebadob/rauthy/issues/166#issuecomment-1891849586), I am thinking about implementing this after the leftover mandatory to become fully OIDC compliant things are done.
I got this as a reminder on the internal TODO list. I am not sure yet how I would want to implement protection against getting spammed with open registration, but I am thinking about a solution.

This however might take some time, since I need to do other things first.

**Comment by sebadob** [comment]
As an update for this issue: I am planning to add dynamic client registration for the next release v0.21.  
This will actually be the last bigger feature to implement properly before the release.

**Comment by sebadob** [comment]
I merged everything needed to make dynamic client registration work.  
I have a few smaller things left on my TODO. When these are done, I will release the next version, which then will implement everything.
